### ISSUE TITLE
testng executes the @BeforeClass and every @Test at a different thread

### ISSUE THREAD

**Issue by yardening2** [testng-team/testng#1333]
### TestNG Version
6.10

### Expected behavior
The whole `<test>name</test>` should run on one thread

### Actual behavior
each @Test inside the class is executed on a different thread

### Is the issue reproductible on runner?

- [ ] Shell
- [ ] Maven
- [ ] Gradle
- [ ] Ant
- [x] Eclipse
- [ ] IntelliJ
- [ ] NetBeans

### Test case sample

So this is the test class. 1 `@BeforeClass` and 2 `@Test`:

```java
public class SpacerDesignTest extends SpacerTestLogic {
	
	@BeforeClass
	public void init(){
		createSite();
		super.init();
		addWidgetByAPI(WidgetType.Spacer);
	}
	
	@Test()
	public void changeSpacerHeight(){
		String height = "150px";
		openDesignWindow();
		changeWidgetHeight(height);
		closeEditWindow();
		Assert.assertEquals(getWidgetHeightFromEmulator(), height, "button width shoud be " + height + " but it's " + getWidgetHeightFromEmulator());
	}
	
	@Test()
	public void changeSpacerHeight1(){
		String height = "150px";
		openDesignWindow();
		changeWidgetHeight(height);
		closeEditWindow();
		Assert.assertEquals(getWidgetHeightFromEmulator(), height, "button width shoud be " + height + " but it's " + getWidgetHeightFromEmulator());
	}

}
```

This is the xml with 3 `<test>` each with the previous class:

```xml
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
<suite name="D1 Tests" verbose="0" parallel="tests" thread-count="10" time-out="7200000">

	<listeners>
    	<listener class-name="com.duda.webdriver.core.SuiteListener" />
    	<listener class-name="com.duda.webdriver.core.TestNGListener" />
	</listeners> 
	
	 <test name="test1">
  	<classes>
  		<class name="com.duda.webdriver.elduderino.tests.widgets.spacer.SpacerDesignTest"></class>
  	</classes>
  	</test>
  	
  	<test name="test2">
  	<classes>
  		<class name="com.duda.webdriver.elduderino.tests.widgets.spacer.SpacerDesignTest"></class>
  	</classes>
  	</test>
  	
  	<test name="test3">
  	<classes>
  		<class name="com.duda.webdriver.elduderino.tests.widgets.spacer.SpacerDesignTest"></class>
  	</classes>
  	</test>
</suite>
```

This is the log i created with the created threads in every `@Test`.
As you can see there are more threads created than expected by this example:

```
Thread is: 1
Thread is: 1
Thread is: 1
Thread is: 1
Thread is: 1
Thread is: 14
Thread is: 14
Thread is: 15
Thread is: 15
Thread is: 16
Thread is: 16
Thread is: 16
Thread is: 14
Thread is: 15
Thread is: 16
Thread is: 14
Thread is: 15
Thread is: 15
Thread is: 14
Thread is: 16
Thread is: 14
Thread is: 16
Thread is: 15
Thread is: 16
Thread is: 15
Thread is: 14
Thread is: 14
Thread is: 14
Thread is: 14
Thread is: 16
Thread is: 16
Thread is: 15
Thread is: 16
Thread is: 15
Thread is: 15
Thread is: 15
Thread is: 15
Thread is: 29
Thread is: 29
Thread is: 29
Thread is: 15
Thread is: 15
Thread is: 15
Thread is: 15
Thread is: 30
Thread is: 30
Thread is: 30
Thread is: 15
Thread is: 15
Thread is: 15
Thread is: 15
Thread is: 14
Thread is: 15
Thread is: 15
Thread is: 14
Thread is: 31
Thread is: 31
Thread is: 31
Thread is: 14
Thread is: 14
Thread is: 14
Thread is: 14
Thread is: 32
Thread is: 32
Thread is: 32
Thread is: 14
Thread is: 14
Thread is: 14
Thread is: 14
Thread is: 16
Thread is: 16
Thread is: 33
Thread is: 33
Thread is: 14
Thread is: 14
Thread is: 33
Thread is: 16
Thread is: 16
Thread is: 16
Thread is: 16
Thread is: 34
Thread is: 34
Thread is: 34
Thread is: 16
Thread is: 16
Thread is: 16
Thread is: 16
Thread is: 16
Thread is: 16
Thread is: 1
Thread is: 1
Thread is: 1
Thread is: 1
Thread is: 1
Thread is: 1
```

**Comment by juherr** [comment]
Duplicate of #1185

Feel free to reopen if you think it is a different issue.

**Comment by juherr** [comment]
Comment from #1335 

> The issue has nothing to do with dependsOn, it happens with the example test I've attached to issue #1333 And there is no dependsOn there

Ok. Do you think it is a duplicate of #1035 in that case?

**Comment by yardening2** [comment]
@juherr 
My case isn't dealing with dependsOn nor the @BeforeClass is not the issue.
Every @Test in the same class gets executed in a different Thread than the previous @Test before it.

In other words, every @Test creates a new thread id.
You can see it in the thread numberings i added to my sample.

This happend to me only after updating my TestNG version from 6.9.10 to 6.10.
It affects my logging mechanisms because it relies on every class is a single thread, on all of its @Tests 

I think when in the testng.xml I specify  parallel="tests" not every @Test should create a new thread.

**Comment by yardening2** [comment]
@juherr 
Ok, so i found the source to this behavior.

Review this example:

```java
public class Test{
	
	@BeforeClass
	public void init(){
		System.out.println(Thread.currentThread().getId());
	}
	
	@Test
	public void test1(){
		System.out.println(Thread.currentThread().getId());
	}
	
	@Test
	public void test2(){
		System.out.println(Thread.currentThread().getId());
	}

}
```

```xml
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
<suite name="D1 Tests" verbose="0" parallel="tests" thread-count="10" time-out="7200000">

	<listeners>
    	<listener class-name="com.duda.webdriver.core.SuiteListener" />
    	<listener class-name="com.duda.webdriver.core.TestNGListener" />
	</listeners> 
	
	 <test name="test1">
  	<classes>
  		<class name="Test"></class>
  	</classes>
  	</test>
  	
  	<test name="test2">
  	<classes>
  		<class name="Test"></class>
  	</classes>
  	</test>
  	
  	<test name="test3">
  	<classes>
  		<class name="Test"></class>
  	</classes>
  	</test>
</suite>
```

I get a different thread id for every method although each <test> tag should be 1 thread.

The reason is the `time-out="7200000"`. When I remove this parameter the tests exectute as expected.

Very strange behavior. Any idea why this happens?

**Comment by juherr** [comment]
@yardening2 Sure! The timeout feature is done by a thread pool on its own threads.

I'm pretty sure you issue, or a similar one, is already registered: https://github.com/cbeust/testng/issues?q=is%3Aopen+is%3Aissue+label%3A%22Feature%3A+timeout%22

**Comment by calottery** [comment]
Hi,
I am a newbie here.  I have a related question. 

In my case, the @BeforeClass method is executed before every test in the class.  The testNG xml file looks like below.  I am trying testing on one device for now - but plan is to test on multiple devices at the same time.

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
<suite thread-count="2" configfailurepolicy="continue" name="Parallel Mobile Tests" parallel="tests" preserve-order="true">
  <listeners>
    <listener class-name="listeners.TestListener"/>
    <listener class-name="listeners.CustomReportListener"/>
  </listeners>
  <test name="IPHONE_6SPLUS : Parallel Tests">
    <parameter name="deviceOS" value="iOS"/>
    <parameter name="device" value="IPHONE_6SPLUS"/>
    <classes>
      <!-- <class name="tests.Install"/> -->
      <class name="tests.Class01"/>
      <class name="tests.Class02"/>
      <class name="tests.Class03"/>
      <class name="tests.Class04"/>
      <class name="tests.Class05"/>
     </classes>
  </test> <!-- IPHONE_6SPLUS : Parallel Tests -->
</suite> <!-- Parallel Mobile Tests -->
```

**Comment by selenUser** [comment]
@juherr there will be a fix for this bug soon ?
when running tests in parallel as methods and had a setup on beforeClass, some test not getting started by the "beforeclass".
thanks

**Comment by krmahadevan** [comment]
@selenUser - Does your scenario involve timeouts ? If not, then please try reproducing this issue using TestNG `7.0.0-beta3` and share a sample reproducible test, if the issue still exists.

**Comment by JuliaIskra** [comment]
as well as @BedoreMethod and @AfterMethod

**Comment by juherr** [comment]
AI triage verdict: **still relevant**.

The current `master` still contains the exact cause: `MethodInvocationHelper.invokeWithTimeout()` only uses the in-place timeout path when the current thread is a TestNG thread **and** the current `<test>` is not `ParallelMode.TESTS`. A suite-level `time-out` is inherited by every test/configuration method, so under `parallel="tests"` each timed method is sent to a fresh single-thread executor and loses the documented one-thread-per-`<test>` affinity.

The earlier suggestions that this was #1185 or #1035 do not match the final reproducer: #1185 is about `dependsOnMethods` with `parallel="classes"`, and #1035 is about factory/`parallel="instances"`. The relevant history is #811, whose timeout fix introduced the special-case premise, and #2009, which later made the in-place path actually interrupt timed methods.

Open PR #3451 directly implements #1333. It removes the now-obsolete `parallel="tests"` exclusion and adds `test.thread.issue1333.Issue1333Test`, covering three `<test>` tags with a suite timeout and a no-timeout control. The PR reports the regression test as red on unmodified production code (3 threads per `<test>`, 9 total) and green with the change, and also strengthens neighboring timeout coverage for #811/#2009. The PR is currently open and unmerged, so the issue should remain open until merge.

I did **not** execute the historical reproducer during this triage; the current behavior is confirmed by the direct code path and by the regression run documented in PR #3451.

Next step: review and merge #3451 after its compatibility trade-off for uninterruptible timed methods under `parallel="tests"` is accepted; then verify the merged regression/timeout suite and close this issue.
